config/v1/types_monitoring.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=monitoring,scope=Cluster
+// +kubebuilder:subresource:status
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/1929
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:internal
+// Monitoring is the Schema for the clustermonitoringoperators API
+type Monitoring struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	// +required
+	Spec MonitoringSpec `json:"spec,omitempty"`
+	// status holds observed values from the cluster. They may not be overridden.
+	// +optional
+	Status MonitoringStatus `json:"status,omitempty"`
+}
+
+// MonitoringOperatorStatus defines the observed state of MonitoringOperator
+type MonitoringStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:internal
+type MonitoringList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Monitoring `json:"items"`
+}
+
+// MonitoringOperatorSpec defines the desired state of MonitoringOperator
+type MonitoringSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// `UserWorkload` enables monitoring for user-defined projects.
+	// +kubebuilder:validation:Enum:="Enabled";"Disabled";
+	// +kubebuilder:default:="Disabled"
+	// +kubebuilder:validation:Optional
+	// +default="Disabled"
+	// +optional
+
+	UserWorkload UserWorkload `json:"enableUserWorkload,omitempty"`
```

Comment:
We could have a top-level field to group settings pertaining to the user-defined monitoring stack which are controlled by the cluster admins. We don't have right now but might in the future (for instance for cross-namespace rules).

````
userWorkload:
  type: Enabled|Disabled
  foo: ...
````

Reply:
+1 for this. We have used this pattern in the configmap already.

Reply:
@jan--f I'm not very happy with `type` though. Would you have a better idea?

Reply:
Discussed offline with @marioferh, the status subresource could be used to expose the different route URLs (e.g. Thanos Q, Prometheus, Alertmanager, ...).

Reply:
Since we want the option to specify things beyond `enabled|disabled` maybe we call the field `Config`. Initially this can have values `Disabled|SingleNamespace`. This would leave the option to add additional "modes' of UWM later. Like so maybe?
```
...
    UserWorkload UserWorkloadConfig
...

type UserWorkloadConfig struct {
    Config UWMMode
}
```
Certainly not ideal naming either but maybe a bit better then `type`?

Reply:
Something like this better?
`
// MonitoringOperatorSpec defines the desired state of MonitoringOperator
type ClusterMonitoringSpec struct {
	// `UserWorkload` enables monitoring for user-defined projects.
	// +kubebuilder:validation:Enum:="Enabled";"Disabled";
	// +kubebuilder:default:="Disabled"
	// +kubebuilder:validation:Optional
	// +default="Disabled"
	// +optional

	UserWorkload UserWorkload `json:"enableUserWorkload,omitempty"`
}

// UserWorkloadConfig is an enumeration that enable or disable monitoring of user-defined projects.
type UserWorkload struct {
	Config UserWorkloadMode
	// To add more UWM config
}

type UserWorkloadMode string

var (
	// UserWorkloadEnabled enables monitoring for user-defined projects.
	UserWorkloadEnabled UserWorkloadMode = "Enabled"
	// UserWorkloadDisabled disables monitoring for user-defined projects.
	UserWorkloadDisabled UserWorkloadMode = "Disabled"
)
`

Reply:
Hmm I'd go with Jan's suggestion here and avoid having a type for enabled|disabled. That would also allow us to have a field for future enhancing of UWM modes. I think that'd work Mario. Just my .02 is to leave this really well defined in either enhancement or comments or the future us would totally forget.